### ISSUE TITLE
RegisterForActivityResult does not work

### ISSUE THREAD

**Issue by dimonovdd** [dotnet/maui#14037]
### Description

`RegisterForActivityResult` does not work correctly with `MauiAppCompatActivity`. 

### Steps to Reproduce

1. Clone the repository
2. Open `AndroidApp1.sln`
3. Enable `Don't keep activities` on Device\Simulator
4. Add breakpoints in two projects in methods `OnActivityResult`
5. Run `AndroidApp1`
6. Touch `Get result`
5. Run `MauiApp1`
6. Touch `Get result`

### Link to public reproduction project repository

https://github.com/dimonovdd/RegisterForActivityResult.MAUI

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Any

**Comment by dimonovdd** [comment]
@PureWeen Hi

Do you have any ideas? These lines have been changed in #1241

**Comment by dimonovdd** [comment]
Backlog? @jsuarezruiz Do you understand the seriousness of this problem? Are we following the path of Xamarin.Froms? If an issue is difficult, then you simply do not have the desire to solve it for several years?


@jonathanpeppers Hi, You are strong in Android development. Do you have any ideas? How can we disable fragment restore without disabling restore of other things?

**Comment by dimonovdd** [comment]
> @dimonovdd it looks like you are supposed to override:

I didn't understand the meaning of it
https://github.com/dotnet/maui/discussions/14039

**Comment by jonathanpeppers** [comment]
I think you override it and return `true`, will prevent `savedInstanceState` from being modified?

**Comment by dimonovdd** [comment]
> will prevent savedInstanceState from being modified?

Yes, but the app will crash. And there is no documentation of what else needs to be changed.

**Comment by jonathanpeppers** [comment]
What is the error message?

**Comment by dimonovdd** [comment]
A screenshot was attached in the [discussion](https://github.com/dotnet/maui/discussions/14039). 
This exception is not only related to `ShellItemRenderer`. Almost nothing has `ctor (IntPtr javaReference, JniHandleOwnership transfer)`.

![](https://user-images.githubusercontent.com/59065470/226067153-9a28df8a-991e-4463-a082-791016994fc2.png)

**Comment by dimonovdd** [comment]
MAUI changed the logic of restoring activity after destroy. These two lines disable default restore of fragments and other Views. But it also disables state restore of other things (RegisterForActivityResult and etc.)

```csharp
savedInstanceState?.Remove("android:support:fragments");
savedInstanceState?.Remove("androidx.lifecycle.BundlableSavedStateRegistry.key");
```

**Comment by PureWeen** [comment]
> @dimonovdd it looks like you are supposed to override:
> 
> ```cs
> protected virtual bool AllowFragmentRestore => false;
> ```
> 
> Did that not work for you?
> 
> @PureWeen can maybe share details on why we need:
> 
> ```
> savedInstanceState?.Remove("android:support:fragments");
> savedInstanceState?.Remove("androidx.lifecycle.BundlableSavedStateRegistry.key");
> ```

The top one has been around forever in Forms https://github.com/xamarin/Xamarin.Forms/pull/246

The second one we added when we bumped up AndroidX because it was causing a similar issue indicated by https://github.com/xamarin/Xamarin.Forms/pull/246

https://github.com/xamarin/Xamarin.Forms/pull/14101/files#diff-89f66fc6d9b2094dcce8863bd3db93823b5ca20d2f2ef588b8e6ffff46fcb0b6R226

Do you happen to know which of these lines is breaking RegisterForActivityResult?

**Comment by dimonovdd** [comment]
> Does that workaround your issue?

No, I changed `MauiAppCompatActivity` code like this and also the exception

![image](https://user-images.githubusercontent.com/59065470/226819659-f0faf2b6-e28f-412a-8fc4-59bef14f3f18.png)

**Comment by PureWeen** [comment]
> > Does that workaround your issue?
> 
> No, I changed `MauiAppCompatActivity` code like this and also the exception
> 
> ![image](https://user-images.githubusercontent.com/59065470/226819659-f0faf2b6-e28f-412a-8fc4-59bef14f3f18.png)

Hmm, yea, it looks like that path broke when we updated AndroidX and then needed to add that other line to account for it.

**Comment by dimonovdd** [comment]
> Hmm, yea, it looks like that path broke when we updated AndroidX and then needed to add that other line to account for it.

I think it will be very painful to tear off this patch. But we definitely have to fix it

**Comment by entdark** [comment]
The issue in the sample is probably in creating `ActivityResultLauncher` before `base.OnCreate`: https://github.com/dimonovdd/RegisterForActivityResult.MAUI/blob/main/MauiApp1/Platforms/Android/MainActivity.cs#L18
Not in MAUI.

**Comment by phunkeler** [comment]
I'm having success with this workaround:

``` csharp
public class MainActivity : MauiAppCompatActivity
{
    protected override bool AllowFragmentRestore => true;

    protected override void OnCreate(Bundle? savedInstanceState)
    {
        ...

        savedInstanceState?
            .GetBundle("androidx.lifecycle.BundlableSavedStateRegistry.key")?
            .Remove("android:support:fragments");

        base.OnCreate(savedInstanceState);
    }
}
```

![14037_api33_workaround](https://github.com/user-attachments/assets/326d0708-0db9-4efd-b63c-beb18f58548e)
